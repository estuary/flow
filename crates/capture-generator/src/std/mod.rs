use async_trait::async_trait;
use http::header::AUTHORIZATION;
use reqwest::{Request, Url};
use schemars::{schema::RootSchema, JsonSchema};
use serde::Deserialize;

use crate::interface::*;

#[derive(Default)]
pub struct BearerTokenAuth {
    config: Option<BearerTokenConfig>,
}
#[derive(Deserialize, JsonSchema)]
pub struct BearerTokenConfig {
    token: String,
}
#[async_trait]
impl Auth for BearerTokenAuth {
    type Config = BearerTokenConfig;

    fn prepare(&mut self, config: BearerTokenConfig) -> Result<(), anyhow::Error> {
        self.config = Some(config);
        Ok(())
    }

    async fn authenticate(&mut self, req: &mut Request) -> Result<(), anyhow::Error> {
        let token = &self.config.as_ref().unwrap().token;
        let headers = req.headers_mut();
        headers.insert(AUTHORIZATION, format!("Bearer {token}").parse().unwrap());
        Ok(())
    }
}

/// A very basic stream that sends a HTTP request to a specific URL.
/// Can be used alongside different paginations to create a stream.
pub struct BasicStream {
    pub key: String,
    pub description: String,
    pub endpoint: String,
    pub method: http::Method,
    pub spec: RootSchema,
}
impl Stream for BasicStream {
    type Config = ();

    fn key(&mut self) -> String {
        self.key.clone()
    }

    fn description(&mut self) -> String {
        self.description.clone()
    }

    fn create_request(&mut self) -> Request {
        let mut request = Request::new(self.method.clone(), Url::parse(&self.endpoint).unwrap());
        request
            .headers_mut()
            .insert("User-Agent", "flow-capture".parse().unwrap());

        let body = request.body_mut();
        *body = Some(vec![].into());

        request
    }

    fn spec(&mut self) -> RootSchema {
        self.spec.clone()
    }
}

pub struct BasicPagination {
    pub page_query_field: String,
    pub current_page_field: String,
    pub max_pages_field: String,
}

#[async_trait]
impl Pagination for BasicPagination {
    type Config = ();

    async fn paginate(
        &mut self,
        req: &mut reqwest::Request,
        last_response: reqwest::Response,
    ) -> Result<bool, anyhow::Error> {
        let last_body = last_response.text().await?;
        println!("last_body {:#?}", last_body);
        let last_body = serde_json::from_str::<serde_json::Value>(&last_body)?;

        let res_current_page = last_body.get(&self.current_page_field).unwrap();
        let res_max_page = last_body.get(&self.max_pages_field).unwrap();

        if res_current_page == res_max_page {
            Ok(false)
        } else {
            let current_page = res_current_page.as_u64().unwrap();
            let next_page = (current_page + 1).to_string();

            req.url_mut()
                .query_pairs_mut()
                .append_pair(&self.page_query_field, &next_page);

            println!("{:#?}", req.url().query());

            Ok(true)
        }
    }
}
