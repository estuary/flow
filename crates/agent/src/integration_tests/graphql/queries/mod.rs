use super::common_setup;
use serde_json::Value;
use std::fs;
use std::path::Path;

#[tokio::test]
#[serial_test::serial]
async fn test_graphql_queries() {
    let (mut harness, alice_user_id, bob_user_id) = common_setup().await;

    // Get the directory containing our query files
    let queries_dir =
        Path::new(env!("CARGO_MANIFEST_DIR")).join("src/integration_tests/graphql/queries");

    // Read all .graphql files in the queries directory
    let entries = fs::read_dir(&queries_dir)
        .expect("Failed to read queries directory")
        .filter_map(|entry| entry.ok())
        .filter(|entry| {
            entry
                .path()
                .extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| ext == "graphql")
                .unwrap_or(false)
        })
        .collect::<Vec<_>>();

    // Sort entries by filename for consistent ordering
    let mut query_files: Vec<_> = entries.into_iter().map(|entry| entry.path()).collect();
    query_files.sort();

    let empty_vars = serde_json::json!({});

    for query_path in query_files {
        let query_content = fs::read_to_string(&query_path)
            .expect(&format!("Failed to read query file: {:?}", query_path));

        let file_stem = query_path
            .file_stem()
            .and_then(|s| s.to_str())
            .expect("Failed to get file stem");

        // Execute query as Alice
        let alice_result: Value = harness
            .execute_graphql_query(alice_user_id, &query_content, &empty_vars)
            .await
            .unwrap_or_else(|e| format!("GraphQL query failed with error: {}", e).into());

        // Execute query as Bob
        let bob_result: Value = harness
            .execute_graphql_query(bob_user_id, &query_content, &empty_vars)
            .await
            .unwrap_or_else(|e| format!("GraphQL query failed with error: {}", e).into());

        // Snapshot the results
        insta::assert_json_snapshot!(format!("alice-{}", file_stem), alice_result);

        insta::assert_json_snapshot!(format!("bob-{}", file_stem), bob_result);
    }
}
