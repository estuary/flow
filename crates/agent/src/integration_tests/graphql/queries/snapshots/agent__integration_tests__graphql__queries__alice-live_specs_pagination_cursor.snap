---
source: crates/agent/src/integration_tests/graphql/queries/mod.rs
expression: alice_result
---
{
  "liveSpecs": {
    "edges": [
      {
        "cursor": "aliceCo/shared/b",
        "node": {
          "catalogName": "aliceCo/shared/b",
          "userCapability": "admin"
        }
      },
      {
        "cursor": "aliceCo/shared/c",
        "node": {
          "catalogName": "aliceCo/shared/c",
          "userCapability": "admin"
        }
      }
    ],
    "pageInfo": {
      "hasNextPage": true,
      "hasPreviousPage": false
    }
  }
}
